File: Calendula/src/androidTest/java/es/usc/citius/servando/calendula/database/DBTest.java
/*
 *    Calendula - An assistant for personal medication management.
 *    Copyright (C) 2014-2018 CiTIUS - University of Santiago de Compostela
 *
 *    Calendula is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this software.  If not, see <http://www.gnu.org/licenses/>.
 */

package es.usc.citius.servando.calendula.database;

import android.support.test.InstrumentationRegistry;
import android.test.InstrumentationTestCase;

import org.joda.time.LocalDate;
import org.joda.time.LocalTime;
import org.junit.Before;
import org.junit.Test;

import es.usc.citius.servando.calendula.CalendulaApp;
import es.usc.citius.servando.calendula.persistence.DailyScheduleItem;
import es.usc.citius.servando.calendula.persistence.Medicine;
import es.usc.citius.servando.calendula.persistence.PickupInfo;
import es.usc.citius.servando.calendula.persistence.Presentation;
import es.usc.citius.servando.calendula.persistence.Routine;
import es.usc.citius.servando.calendula.persistence.Schedule;
import es.usc.citius.servando.calendula.persistence.ScheduleItem;

public class DBTest extends InstrumentationTestCase {

    @Before
    public void setUp() throws Exception {
        super.setUp();
        CalendulaApp.disableReceivers = true;
        injectInstrumentation(InstrumentationRegistry.getInstrumentation());
        DB.init(getInstrumentation().getContext());
        DB.dropAndCreateDatabase();
    }

    @Test
    public void testDaoSave() throws Exception {
        Routine r = new Routine(new LocalTime(0, 0), "Test");
        Medicine m = new Medicine("TestMed", Presentation.CAPSULES);
        Schedule s = new Schedule(m);
        ScheduleItem i = new ScheduleItem(s, r);
        DailyScheduleItem d = new DailyScheduleItem(i);
        d.setTakenToday(true);

        PickupInfo pk = new PickupInfo();
        pk.setFrom(LocalDate.parse("2015-01-01"));
        pk.setTo(LocalDate.parse("2015-02-05"));
        pk.setMedicine(m);

        // save some stuff
        DB.routines().save(r);
        DB.medicines().save(m);
        DB.schedules().save(s);
        DB.scheduleItems().save(i);
        DB.dailyScheduleItems().save(d);
        DB.pickups().save(pk);

        // verify id property created
        assertNotNull(r.getId());
        assertNotNull(m.getId());
        assertNotNull(s.getId());
        assertNotNull(i.getId());
        assertNotNull(d.getId());
        assertNotNull(pk.getId());

        assertEquals(DB.pickups().findByMedicine(m).get(0).getFrom(), LocalDate.parse("2015-01-01"));
    }

    @Test
    public void testDaoFind() throws Exception {

        testDaoSave();
        assertEquals(DB.routines().findOneBy(Routine.COLUMN_NAME, "Test").getTime(), new LocalTime(0, 0));
        assertEquals(DB.medicines().findAll().get(0).getName(), "TestMed");
        assertEquals(DB.schedules().findAll().get(0).items().size(), 1);
    }


}

File: Calendula/src/test/java/es/usc/citius/servando/calendula/database/DBTest.kt
/*
 *    Calendula - An assistant for personal medication management.
 *    Copyright (C) 2014-2018 CiTIUS - University of Santiago de Compostela
 *
 *    Calendula is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this software.  If not, see <http://www.gnu.org/licenses/>.
 */

package es.usc.citius.servando.calendula.database

import es.usc.citius.servando.calendula.CalendulaApp
import es.usc.citius.servando.calendula.persistence.*
import org.codehaus.plexus.util.FileUtils
import org.joda.time.LocalDate
import org.joda.time.LocalTime
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import java.io.File
import java.time.Clock


@RunWith(RobolectricTestRunner::class)
class DBTest {

    companion object {
        private const val MED_NAME = "TestMed"
        private const val ROUTINE_NAME = "Test"
        private const val FROM_DATE = "2015-01-01"
        private const val TO_DATE = "2015-02-05"
    }

    @Before
    fun setUp() {
        CalendulaApp.disableReceivers = true
        DB.init(RuntimeEnvironment.application)
        DB.dropAndCreateDatabase()
    }

    @Test
    fun testDaoSave() {
        val r = Routine(LocalTime(0, 0), ROUTINE_NAME)
        val m = Medicine(MED_NAME, Presentation.CAPSULES)
        val s = Schedule(m)
        val i = ScheduleItem(s, r)
        val d = DailyScheduleItem(i)
        d.takenToday = true

        val pk = PickupInfo()
        pk.from = LocalDate.parse(FROM_DATE)
        pk.to = LocalDate.parse(TO_DATE)
        pk.medicine = m

        // save some stuff
        DB.routines().save(r)
        DB.medicines().save(m)
        DB.schedules().save(s)
        DB.scheduleItems().save(i)
        DB.dailyScheduleItems().save(d)
        DB.pickups().save(pk)

        // verify id property created
        assertNotNull(r.id)
        assertNotNull(m.id)
        assertNotNull(s.id)
        assertNotNull(i.id)
        assertNotNull(d.id)
        assertNotNull(pk.id)

        assertEquals(DB.pickups().findByMedicine(m)[0].from, LocalDate.parse(FROM_DATE))
    }

    @Test
    fun testDaoFind() {

        testDaoSave()
        assertEquals(
            DB.routines().findOneBy(Routine.COLUMN_NAME, ROUTINE_NAME).time,
            LocalTime(0, 0)
        )
        assertEquals(DB.medicines().findAll()[0].name, MED_NAME)
        assertEquals(DB.schedules().findAll()[0].items().size.toLong(), 1)
    }

    /**
     * Abuse test framework to export database into a project folder, allowing further analysis of database schema
     */
    @Test
    fun exportDatabase() {
        val db = DB.helper().writableDatabase;
        DB.helper().close();
        val destination = File("../generated/db")
        destination.mkdirs()
        System.out.println(String.format("Exporting application database %s to %s", db.path, destination.absolutePath))
        FileUtils.copyFileToDirectory(File(db.path), destination)
        assertNotNull(db.path);
    }
}
